### DISCUSSION TITLE
Hide homespan.poll() under the hood?

### DISCUSSION THREAD

**Discussion by danilkorotkov** [HomeSpan/HomeSpan#304] [Ideas]
https://github.com/HomeSpan/HomeSpan/discussions/262
is it so necessary to keep poll in loop?

**Comment by HomeSpan** [comment]
Probably not - it is something I will explore in the future.  Want to test will the single-core chips as well.  Also want to see how it can be integrated to automatically start at end of setup() but before loop().  I like to keep backwards compatibility with all sketches and can deprecate the poll() function while setting it to NOP.  Setting poll as a task can be done with something like adding homeSpan.start(), at the end of setup(), but ideally the code would do this automatically.

The Arduino-ESP32 library has various hooks in place to call init().  I should be able to find an appropriate one to use to setup the poll task, making the entire thing invisible to the user.

**Comment by HomeSpan** [comment]
The Arduino-ESP32 file *main.c* calls initArduino() and then creates a task for the Arduino loop() function.  The function initArduino() is found in *esp32-hal-misc.c*, and after performing some configuration tasks (including a not-so-useful ROLLBACK routine), it ends by calling init() *and* then initVariant().  The latter is a weakly-defined NOP function.  If this function is *not* used by any of the board variants (I do not think it is), then it would be the perfect place to create the poll() task, where it would be invisible to the user.  Will give it a try this weekend.

**Reply by danilkorotkov** [reply]
Please, do not forget to create  core selection method or land poll() automatically to free core for 2core chips.

**Comment by HomeSpan** [comment]
Done.  There was unfortunately no "hook" into which I could automatically call xTask from within the library.  The init() and initVariant() functions I noted above occur before the Arduino setup() function, so I could not use them.

Instead, I added a new command homeSpan.start() that users can place at the end of the Arduino setup() function.  This starts the polling task in the background on core 0 for multi-core processors.  Even on a single-core processor this is useful since it frees up the loop() function for any user-defined code without worrying about blocking, or being blocked by, homeSpan.

Error checking throws an error if homeSpan.poll() is called if you've already used homeSpan.start() - only one method or the other should be used.

Code is updated in dev branch - give it a try (will be adding to documentation on dev branch as well shortly)

**Comment by HomeSpan** [comment]
Update:  renamed `homeSpan.start()` to `homeSpan.autoPoll()`.  Adding documentation to API reference.  Changes are in dev branch.

**Reply by danilkorotkov** [reply]
Now I’m porting tuya thermostat (esp8266) to HomeKit. I will test later
